File: db4oj/core/src/com/db4o/IxField.java
/* Copyright (C) 2004   db4objects Inc.   http://www.db4o.com */

package com.db4o;

import com.db4o.foundation.*;

/**
 * 
 * @exclude
 */
public class IxField {

    static final int   MAX_LEAVES = 3;

    static private int i_version;

    final YapField     i_field;

    final MetaIndex    i_metaIndex;

    IxFieldTransaction i_globalIndex;

    Collection4        i_transactionIndices;

    IxFileRangeReader  i_fileRangeReader;

    IxField(Transaction a_systemTrans, YapField a_field, MetaIndex a_metaIndex) {
        i_metaIndex = a_metaIndex;
        i_field = a_field;
        i_globalIndex = new IxFieldTransaction(a_systemTrans, this);
        createGlobalFileRange();
    }

    IxFieldTransaction dirtyFieldTransaction(Transaction a_trans) {
        IxFieldTransaction ift = new IxFieldTransaction(a_trans, this);
        if (i_transactionIndices == null) {
            i_transactionIndices = new Collection4();
        } else {
            IxFieldTransaction iftExisting = (IxFieldTransaction) i_transactionIndices
                    .get(ift);
            if (iftExisting != null) {
                return iftExisting;
            }
        }
        a_trans.addDirtyFieldIndex(ift);
        ift.setRoot(Tree.deepClone(i_globalIndex.getRoot(), ift));
        ift.i_version = ++i_version;
        i_transactionIndices.add(ift);
        return ift;
    }

    IxFieldTransaction getFieldTransaction(Transaction a_trans) {
        if (i_transactionIndices != null) {
            IxFieldTransaction ift = new IxFieldTransaction(a_trans, this);
            ift = (IxFieldTransaction) i_transactionIndices.get(ift);
            if (ift != null) {
                return ift;
            }
        }
        return i_globalIndex;
    }
    
// Debug index tree depth    
    
//    void debug(IxFieldTransaction a_ft){
//        System.out.println("+++  IxField commit debug begin");
//        Tree t1 = i_globalIndex.getRoot();
//        if (t1 != null){
//            System.out.println("i_globalIndex");
//            t1.debugDepth();
//        }
//        if(a_ft != null){
//            Tree t2 = a_ft.getRoot();
//            if (t2 != null){
//                System.out.println("a_ft");
//                t2.debugDepth();
//            }
//        }
//        System.out.println("---  IxField commit debug complete");
//    }

    void commit(IxFieldTransaction a_ft) {
        i_transactionIndices.remove(a_ft);

        i_globalIndex.merge(a_ft);

        int leaves = i_globalIndex.countLeaves();

        // TODO: Use more intelligent heuristic here to
        // calculate when to flush the global index
        // boolean createNewFileRange = leaves > MAX_LEAVES;
        boolean createNewFileRange = true;

        if (createNewFileRange) {
            final Transaction trans = i_globalIndex.i_trans;

            int[] free = new int[] { i_metaIndex.indexAddress,
                    i_metaIndex.indexLength, i_metaIndex.patchAddress,
                    i_metaIndex.patchLength};

            Tree root = i_globalIndex.getRoot();
            final YapDataType handler = i_field.getHandler();

            final int lengthPerEntry = handler.linkLength()
                    + YapConst.YAPINT_LENGTH;

            i_metaIndex.indexEntries = root == null ? 0 : root.size();
            i_metaIndex.indexLength = i_metaIndex.indexEntries * lengthPerEntry;
            i_metaIndex.indexAddress = ((YapFile) trans.i_stream)
                    .getSlot(i_metaIndex.indexLength);
            i_metaIndex.patchEntries = 0;
            i_metaIndex.patchAddress = 0;
            i_metaIndex.patchLength = 0;
            trans.i_stream.setInternal(trans, i_metaIndex, 1, false);

            final YapWriter writer = new YapWriter(trans,
                    i_metaIndex.indexAddress, lengthPerEntry);
            if (root != null) {
                root.traverse(new Visitor4() {

                    public void visit(Object a_object) {
                        ((IxTree) a_object).write(handler, writer);
                    }
                });
            }
            IxFileRange newFileRange = createGlobalFileRange();

            Iterator4 i = i_transactionIndices.iterator();
            while (i.hasNext()) {
                final IxFieldTransaction ft = (IxFieldTransaction) i.next();
                Tree clonedTree = newFileRange;
                if (clonedTree != null) {
                    clonedTree = clonedTree.deepClone(ft);
                }
                final Tree[] tree = { clonedTree};
                ft.getRoot().traverseFromLeaves((new Visitor4() {
                    
                    public void visit(Object a_object) {
                        IxTree ixTree = (IxTree) a_object;
                        if (ixTree.i_version == ft.i_version) {
                            if (!(ixTree instanceof IxFileRange)) {
                                ixTree.beginMerge();
                                tree[0] = tree[0].add(ixTree);
                            }
                        }
                    }
                }));
                ft.setRoot(tree[0]);
            }

            if (free[0] > 0) {
                trans.i_file.free(free[0], free[1]);
            }
            if (free[2] > 0) {
                trans.i_file.free(free[2], free[3]);
            }
        } else {
            Iterator4 i = i_transactionIndices.iterator();
            while (i.hasNext()) {
                ((IxFieldTransaction) i.next()).merge(a_ft);
            }
        }
    }

    private IxFileRange createGlobalFileRange() {
        IxFileRange fr = null;
        if (i_metaIndex.indexEntries > 0) {
            fr = new IxFileRange(i_globalIndex,
                    i_metaIndex.indexAddress, 0, i_metaIndex.indexEntries);
        }
        i_globalIndex.setRoot(fr);
        return fr;
    }

    void rollback(IxFieldTransaction a_ft) {
        i_transactionIndices.remove(a_ft);
    }

    IxFileRangeReader fileRangeReader() {
        if (i_fileRangeReader == null) {
            i_fileRangeReader = new IxFileRangeReader(i_field.getHandler());
        }
        return i_fileRangeReader;
    }

    public String toString() {
        StringBuffer sb = new StringBuffer();
        sb.append("IxField  " + System.identityHashCode(this));
        if (i_globalIndex != null) {
            sb.append("\n  Global \n   ");
            sb.append(i_globalIndex.toString());
        } else {
            sb.append("\n  no global index \n   ");
        }
        if (i_transactionIndices != null) {
            Iterator4 i = i_transactionIndices.iterator();
            while (i.hasNext()) {
                sb.append("\n");
                sb.append(i.next().toString());
            }
        }
        return sb.toString();
    }

}

File: db4on/core/src/core/com/db4o/IxField.cs

namespace com.db4o
{
	/// <exclude></exclude>
	public class IxField
	{
		internal const int MAX_LEAVES = 3;

		private static int i_version;

		internal readonly com.db4o.YapField i_field;

		internal readonly com.db4o.MetaIndex i_metaIndex;

		internal com.db4o.IxFieldTransaction i_globalIndex;

		internal com.db4o.foundation.Collection4 i_transactionIndices;

		internal com.db4o.IxFileRangeReader i_fileRangeReader;

		internal IxField(com.db4o.Transaction a_systemTrans, com.db4o.YapField a_field, com.db4o.MetaIndex
			 a_metaIndex)
		{
			i_metaIndex = a_metaIndex;
			i_field = a_field;
			i_globalIndex = new com.db4o.IxFieldTransaction(a_systemTrans, this);
			createGlobalFileRange();
		}

		internal virtual com.db4o.IxFieldTransaction dirtyFieldTransaction(com.db4o.Transaction
			 a_trans)
		{
			com.db4o.IxFieldTransaction ift = new com.db4o.IxFieldTransaction(a_trans, this);
			if (i_transactionIndices == null)
			{
				i_transactionIndices = new com.db4o.foundation.Collection4();
			}
			else
			{
				com.db4o.IxFieldTransaction iftExisting = (com.db4o.IxFieldTransaction)i_transactionIndices
					.get(ift);
				if (iftExisting != null)
				{
					return iftExisting;
				}
			}
			a_trans.addDirtyFieldIndex(ift);
			ift.setRoot(com.db4o.Tree.deepClone(i_globalIndex.getRoot(), ift));
			ift.i_version = ++i_version;
			i_transactionIndices.add(ift);
			return ift;
		}

		internal virtual com.db4o.IxFieldTransaction getFieldTransaction(com.db4o.Transaction
			 a_trans)
		{
			if (i_transactionIndices != null)
			{
				com.db4o.IxFieldTransaction ift = new com.db4o.IxFieldTransaction(a_trans, this);
				ift = (com.db4o.IxFieldTransaction)i_transactionIndices.get(ift);
				if (ift != null)
				{
					return ift;
				}
			}
			return i_globalIndex;
		}

		internal virtual void commit(com.db4o.IxFieldTransaction a_ft)
		{
			i_transactionIndices.remove(a_ft);
			i_globalIndex.merge(a_ft);
			int leaves = i_globalIndex.countLeaves();
			bool createNewFileRange = true;
			if (createNewFileRange)
			{
				com.db4o.Transaction trans = i_globalIndex.i_trans;
				int[] free = new int[] { i_metaIndex.indexAddress, i_metaIndex.indexLength, i_metaIndex
					.patchAddress, i_metaIndex.patchLength };
				com.db4o.Tree root = i_globalIndex.getRoot();
				com.db4o.YapDataType handler = i_field.getHandler();
				int lengthPerEntry = handler.linkLength() + com.db4o.YapConst.YAPINT_LENGTH;
				i_metaIndex.indexEntries = root == null ? 0 : root.size();
				i_metaIndex.indexLength = i_metaIndex.indexEntries * lengthPerEntry;
				i_metaIndex.indexAddress = ((com.db4o.YapFile)trans.i_stream).getSlot(i_metaIndex
					.indexLength);
				i_metaIndex.patchEntries = 0;
				i_metaIndex.patchAddress = 0;
				i_metaIndex.patchLength = 0;
				trans.i_stream.setInternal(trans, i_metaIndex, 1, false);
				com.db4o.YapWriter writer = new com.db4o.YapWriter(trans, i_metaIndex.indexAddress
					, lengthPerEntry);
				if (root != null)
				{
					root.traverse(new _AnonymousInnerClass119(this, handler, writer));
				}
				com.db4o.IxFileRange newFileRange = createGlobalFileRange();
				com.db4o.foundation.Iterator4 i = i_transactionIndices.iterator();
				while (i.hasNext())
				{
					com.db4o.IxFieldTransaction ft = (com.db4o.IxFieldTransaction)i.next();
					com.db4o.Tree clonedTree = newFileRange;
					if (clonedTree != null)
					{
						clonedTree = clonedTree.deepClone(ft);
					}
					com.db4o.Tree[] tree = { clonedTree };
					ft.getRoot().traverseFromLeaves((new _AnonymousInnerClass136(this, ft, tree)));
					ft.setRoot(tree[0]);
				}
				if (free[0] > 0)
				{
					trans.i_file.free(free[0], free[1]);
				}
				if (free[2] > 0)
				{
					trans.i_file.free(free[2], free[3]);
				}
			}
			else
			{
				com.db4o.foundation.Iterator4 i = i_transactionIndices.iterator();
				while (i.hasNext())
				{
					((com.db4o.IxFieldTransaction)i.next()).merge(a_ft);
				}
			}
		}

		private sealed class _AnonymousInnerClass119 : com.db4o.foundation.Visitor4
		{
			public _AnonymousInnerClass119(IxField _enclosing, com.db4o.YapDataType handler, 
				com.db4o.YapWriter writer)
			{
				this._enclosing = _enclosing;
				this.handler = handler;
				this.writer = writer;
			}

			public void visit(object a_object)
			{
				((com.db4o.IxTree)a_object).write(handler, writer);
			}

			private readonly IxField _enclosing;

			private readonly com.db4o.YapDataType handler;

			private readonly com.db4o.YapWriter writer;
		}

		private sealed class _AnonymousInnerClass136 : com.db4o.foundation.Visitor4
		{
			public _AnonymousInnerClass136(IxField _enclosing, com.db4o.IxFieldTransaction ft
				, com.db4o.Tree[] tree)
			{
				this._enclosing = _enclosing;
				this.ft = ft;
				this.tree = tree;
			}

			public void visit(object a_object)
			{
				com.db4o.IxTree ixTree = (com.db4o.IxTree)a_object;
				if (ixTree.i_version == ft.i_version)
				{
					if (!(ixTree is com.db4o.IxFileRange))
					{
						ixTree.beginMerge();
						tree[0] = tree[0].add(ixTree);
					}
				}
			}

			private readonly IxField _enclosing;

			private readonly com.db4o.IxFieldTransaction ft;

			private readonly com.db4o.Tree[] tree;
		}

		private com.db4o.IxFileRange createGlobalFileRange()
		{
			com.db4o.IxFileRange fr = null;
			if (i_metaIndex.indexEntries > 0)
			{
				fr = new com.db4o.IxFileRange(i_globalIndex, i_metaIndex.indexAddress, 0, i_metaIndex
					.indexEntries);
			}
			i_globalIndex.setRoot(fr);
			return fr;
		}

		internal virtual void rollback(com.db4o.IxFieldTransaction a_ft)
		{
			i_transactionIndices.remove(a_ft);
		}

		internal virtual com.db4o.IxFileRangeReader fileRangeReader()
		{
			if (i_fileRangeReader == null)
			{
				i_fileRangeReader = new com.db4o.IxFileRangeReader(i_field.getHandler());
			}
			return i_fileRangeReader;
		}

		public override string ToString()
		{
			j4o.lang.StringBuffer sb = new j4o.lang.StringBuffer();
			sb.append("IxField  " + j4o.lang.JavaSystem.identityHashCode(this));
			if (i_globalIndex != null)
			{
				sb.append("\n  Global \n   ");
				sb.append(i_globalIndex.ToString());
			}
			else
			{
				sb.append("\n  no global index \n   ");
			}
			if (i_transactionIndices != null)
			{
				com.db4o.foundation.Iterator4 i = i_transactionIndices.iterator();
				while (i.hasNext())
				{
					sb.append("\n");
					sb.append(i.next().ToString());
				}
			}
			return sb.ToString();
		}
	}
}
